numba/cuda/cudaimpl.py

```diff
@@ -492,6 +492,104 @@ def ptx_fp16_binary(context, builder, sig, args):
+                           "rcp.approx.ftz.f32 $0, $1;",
+                           "=f,f")
+    arg2_rcp = builder.call(rcp_asm, [arg2_fp32])
+
+    # Multiply by reciprocal of arg2
+    fmul = builder.fmul(arg1_fp32, arg2_rcp)
+
+    #Convert from fp32 to fp16
+    cvt2fp16_fnty = ir.FunctionType(ir.IntType(16), [ir.FloatType()])
+    cvt2fp16_asm = ir.InlineAsm(cvt2fp16_fnty,
+                                "cvt.rn.f16.f32 $0, $1;",
+                                "=h,f")
+    fp16_div = builder.call(cvt2fp16_asm, [fmul])
+
+    and_cnst = context.get_constant(types.int16, 32767)
+    and_op = builder.and_(fp16_div, and_cnst)
+
+    #temp_reg = get_temp_predicate_register()
+    temp_reg = '__$$temp3'
+    reg_pred_fnty = ir.FunctionType(ir.VoidType(),[])
+    reg_pred_asm = ir.InlineAsm(reg_pred_fnty,
+                                f"{{ .reg .pred {temp_reg};",
+                                "")
+    _ = builder.call(reg_pred_asm, [])
+
+    setp_fnty = ir.FunctionType(ir.VoidType(),
+                                [ir.IntType(16), ir.IntType(16)])
+    setp_cnst = context.get_constant(types.int16, 143)
+    setp_asm = ir.InlineAsm(setp_fnty,
+                            f"setp.lt.f16 {temp_reg}, $0, $1;",
+                            'h,h')
+    _ = builder.call(setp_asm, [and_op, setp_cnst])
+
+    selp_fnty = ir.FunctionType(ir.IntType(16),[])
+
+    selp_asm = ir.InlineAsm(selp_fnty,
+                            f"selp.u16 $0, 1, 0, {temp_reg};}}",
+                            '=h')
+    selp = builder.call(selp_asm, [])
+
+    zero = context.get_constant(types.int16, 0)
+    cmp1 = builder.icmp_unsigned("==",
+                                 builder.bitcast(selp, ir.IntType(16)),
+                                 zero)
+    cmp2 = builder.icmp_unsigned("==",
+                                 builder.bitcast(and_op, ir.IntType(16)),
+                                 zero)
+    or_op = builder.or_(cmp1, cmp2)
+    or_cond = builder.bitcast(or_op, ir.IntType(1))
+
+    entry_bb = builder.block
+    fall_thru_bb = builder.append_basic_block()
+    true_bb = builder.append_basic_block()
+    builder.cbranch(or_cond, true_bb, fall_thru_bb)
+
+    #Fall thru branch
+    builder.position_at_start(fall_thru_bb)
+
+    float_zero = context.get_constant(types.float32, 0.0)
+    neg_f = builder.fsub(float_zero, arg2_fp32)
+
+    fname = 'llvm.nvvm.fma.rn.f'
+    lmod = builder.module
+    fnty = ir.FunctionType(ir.FloatType(),
+                           (ir.FloatType(),
+                            ir.FloatType(),
+                            ir.FloatType()))
+    fma_func = cgutils.get_or_insert_function(lmod, fnty, fname)
+    fma1 = builder.call(fma_func, [neg_f, fmul, arg2_fp32])
+    fma2 = builder.call(fma_func, [arg2_rcp, fma1, fmul])
+    fp32_to_f16 = builder.call(cvt2fp16_asm, [fma2])
+    builder.branch(true_bb)
+
+    #True branch
+    builder.position_at_start(true_bb)
+    phi_node = builder.phi(ir.IntType(16))
```

Comment:
Does the `phi` need to be a 16 bit int?

Reply:
Yes it needs to be a 16-bit integer type because the Float model for fp16 is implemented as a 16-bit type:

@register_default(types.Float)
class FloatModel(models.PrimitiveModel):
    def __init__(self, dmm, fe_type):
        if fe_type == types.float16:
            be_type = ir.IntType(16)
        elif fe_type == types.float32:
            be_type = ir.FloatType()
        elif fe_type == types.float64:
            be_type = ir.DoubleType()
        else:
            raise NotImplementedError(fe_type)
        super(FloatModel, self).__init__(dmm, fe_type, be_type)